### PR TITLE
Optimize CommandRunner and Commands

### PR DESCRIPTION
**Pull request by paulbalandan**
**Checklist:**
- [x] Securely signed commits
- [x] Component(s) with PHPdocs
- [x] Unit testing, with >80% coverage
- [x] Conforms to style guide

### CODE REVIEW

**Comment by paulbalandan** [comment]
@MGatner this optimization made all test cases of `CommandRunnerTest` to be all under 0.50 s.

File: system/CLI/CommandRunner.php [hunk 1]
```diff
@@ -71,13 +67,9 @@ public function _remap($method, ...$params)
 	public function index(array $params)
 	{
 		$command = array_shift($params);
+		$command = $command ?? 'list';
```
Comment (MGatner) [inline system/CLI/CommandRunner.php:70]:
```suggestion
 		$command = array_shift($params) ?? 'list';
```

File: system/CLI/Commands.php [hunk 2]
```diff
@@ -77,39 +74,38 @@ public function run(string $command, array $params)
...
 	public function discoverCommands()
 	{
-		if (! empty($this->commands))
+		if ($this->commands !== [])
 		{
 			return;
 		}
 
-		$files = service('locator')->listFiles('Commands/');
+		$locator = Services::locator();
```
Comment (MGatner) [inline system/CLI/Commands.php:93]: Lonnie wants us to start using the helper function. 
```suggestion
		$locator = service('locator');
```

File: tests/system/CLI/CommandRunnerTest.php [hunk 3]
```diff
@@ -1,66 +1,49 @@
...
-	private $stream_filter;
-	protected $env;
-	protected $config;
-	protected $request;
-	protected $response;
-	protected $logger;
-	protected $runner;
+	public static function setUpBeforeClass(): void
+	{
+		self::$logger = Services::logger();
```
Comment (MGatner) [inline tests/system/CLI/CommandRunnerTest.php:23]:
```suggestion
		self::$logger = service('logger');
```

File: tests/system/CLI/CommandRunnerTest.php [hunk 4]
```diff
@@ -1,66 +1,49 @@
...
+		self::$runner = new CommandRunner();
+
+		self::$runner->initController(Services::request(), Services::response(), self::$logger);
```
Comment (MGatner) [inline tests/system/CLI/CommandRunnerTest.php:26]: Service helper

**Review by MGatner — CHANGES_REQUESTED** [review]
Glad you cracked this egg! Moving the discovery to the constructor could cause some people problems, but generally I think it is a better design. Maybe see if we can get another review from @lonnieezell or @samsonasik on that point? Otherwise just a couple of tweaks.

**Comment by lonnieezell** [comment]
I typically prefer not to do things like that in a constructor but in this case I don't see a problem with it. Commands aren't something that are called all of the time, and testing is where that has the tendency to be the biggest issue. But that's why we're moving it now so it looks good to me!

**Comment by paulbalandan** [comment]
> Glad you cracked this egg! Moving the discovery to the constructor could cause some people problems, but generally I think it is a better design. Maybe see if we can get another review from @lonnieezell or @samsonasik on that point? Otherwise just a couple of tweaks.

I have pondered on this and I don't think moving discovery to the constructor would give us problems. Discovery essentially fills up the $commands array prop with the discovered commands. Current design shows that every invoke of run we call discoverCommands which in turn returns early if the $commands prop is already populated. So this prop is never overwritten. Moving the discovery at constructor initialization would be a big gain, as the tests demonstrate.

**Review by samsonasik — APPROVED** [review]

**Review by MGatner — APPROVED** [review]

**Pull request merged**